File: 871-keys-and-rooms/keys-and-rooms.dart
import 'dart:collection';

class Solution {
  bool canVisitAllRooms(List<List<int>> rooms) {
    final visitedRooms = <int>{};
    final queue = Queue<List<int>>();

    visitedRooms.add(0);
  
    queue.addLast(rooms[0]);

    while (queue.isNotEmpty) {
      final room = queue.removeFirst();

      for (final key in room) {
        if (visitedRooms.contains(key)) continue;

        visitedRooms.add(key);

        queue.addLast(rooms[key]);
      }
    }

    return rooms.length == visitedRooms.length;
  }
}

File: 871-keys-and-rooms/keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = {0}
        self.dfs(rooms, 0, visited)
        return len(visited) == len(rooms)

    def dfs(self, rooms, node, visited):
        neighbors = rooms[node]

        for neighbor in neighbors:
            if neighbor in visited:
                continue
            visited.add(neighbor)
            self.dfs(rooms, neighbor, visited)
        